internal/controller/drplacementcontrol_controller.go

```diff
@@ -1311,6 +1313,38 @@ func (r *DRPlacementControlReconciler) updateDRPCStatus(
 	return nil
 }
 
+func (r *DRPlacementControlReconciler) repairReverseProgression(drpc *rmn.DRPlacementControl, log logr.Logger) {
+	// if Phase is changed - no need to manipulate Progression, it can be changed in both directions
+	if r.savedInstanceStatus.Phase != drpc.Status.Phase {
+		return
+	}
+
+	var progressionList []rmn.ProgressionStatus
+
+	switch drpc.Spec.Action {
+	case rmn.ActionFailover:
+		progressionList = failoverProgressionList
+	case rmn.ActionRelocate:
+		progressionList = relocateProgressionList
+	default:
+		progressionList = deployProgressionList
+	}
+
+	currentProgressionIndex := slices.Index(progressionList, r.savedInstanceStatus.Progression)
+	nextProgressionIndex := slices.Index(progressionList, drpc.Status.Progression)
+
+	if currentProgressionIndex == -1 || nextProgressionIndex == -1 {
+		return
+	}
+
+	if nextProgressionIndex < currentProgressionIndex {
+		log.Info(fmt.Sprintf("Repair reverse Progression from '%s' to '%s'",
+			drpc.Status.Progression, r.savedInstanceStatus.Progression))
```

Comment:
Repairing the progression and logging is not the way to go. We should instead not set the wrong progression.

If we are sure we can do this, we can change the place setting the progression to reject value which will make go backward. At that point we can log a debug message about rejecting the progression.

Reply:
@nirs We are talking about this code:
```
ready := d.checkReadiness(failoverCluster)
if !ready {
	d.log.Info("VRGCondition not ready to finish failover")
	d.setProgression(rmn.ProgressionWaitForReadiness)

	return !done, nil
}

```
If there is generation mismatch, we are requeuing, so next reconciliation may or may not solve it.

Reply:
> @nirs We are talking about this code:
> 
> ```
> ready := d.checkReadiness(failoverCluster)
> if !ready {
> 	d.log.Info("VRGCondition not ready to finish failover")
> 	d.setProgression(rmn.ProgressionWaitForReadiness)
> 
> 	return !done, nil
> }
> ```
> 
> If there is generation mismatch, we are requeuing, so next reconciliation may or may not solve it. But every reconcile we are going through several steps, even if we already completed the preferred action.

This is the bug! checkReadyness returns false because the generations do not match. It should fail with an error since it cannot tell if we are ready or not. The resource status wil not be updated until the generations match.